libyul/YulName.h

```diff
@@ -0,0 +1,197 @@
+
+	BuiltinFunction const* discardFunction(YulName _type) const;
+	BuiltinFunction const* equalityFunction(YulName _type) const;
+	BuiltinFunction const* booleanNegationFunction() const;
+
+	BuiltinFunction const* memoryStoreFunction(YulName _type) const;
+	BuiltinFunction const* memoryLoadFunction(YulName _type) const;
+	BuiltinFunction const* storageStoreFunction(YulName _type) const;
+	BuiltinFunction const* storageLoadFunction(YulName _type) const;
+	YulName hashFunction(YulName _type) const;
+
+	/// Tries to find the label in the defined names and returns the corresponding name. If not found, ``emptyName``.
+	YulName nameOfLabel(std::string_view label) const;
+
+	/// Same as nameOfLabel, but restricted to builtins and therefore computationally more efficient.
+	YulName nameOfBuiltin(std::string_view builtin) const;
+
+	/// Same as nameOfLabel, but restricted to types and therefore computationally more efficient.
+	YulName nameOfType(std::string_view type) const;
+
+	/// Whether the name is corresponding to a type.
+	bool isType(YulName _name) const;
+
+	/// Number of types. If the dialect is untyped, there is still one type (the "empty type" type)
+	size_t nTypes() const;
+
+	Dialect const& dialect() const;
+
+	bool isEvmDialect() const;
+
+	/// Generates labels for derived names over the set of _usedNames, respecting a set of _illegal labels.
+	/// This will change the state of all derived names in _usedNames to "not derived" with a label associated to them.
+	void generateLabels(std::set<YulName> _usedNames, std::set<std::string> const& _illegal = {});
+
+	// commented out for the time being, as it requires AST refactoring to be YulName-based
+	// void generateLabels(Block const& _ast, std::set<std::string> const& _illegal = {});
+
+private:
+	bool nameWithinBounds(YulName const _name) const { return _name < m_index; }
+
+	size_t indexOfType(YulName _type) const;
+	BuiltinFunction convertBuiltinFunction(YulName _name, yul::BuiltinFunction const& _builtin) const;
+	BuiltinFunction const* fetchTypedPredefinedFunction(YulName _type, std::vector<std::optional<YulName>> const& _functions) const;
+
+	Dialect const& m_dialect;
+	std::vector<std::tuple<YulName, std::string>> m_dialectTypes;
+	std::map<YulName, BuiltinFunction> m_builtinFunctions;
+
+	struct PredefinedBuiltinFunctions
+	{
+		std::vector<std::optional<YulName>> discardFunctions;
+		std::vector<std::optional<YulName>> equalityFunctions;
+		std::optional<YulName> booleanNegationFunction;
+		std::vector<std::optional<YulName>> memoryStoreFunctions;
+		std::vector<std::optional<YulName>> memoryLoadFunctions;
+		std::vector<std::optional<YulName>> storageStoreFunctions;
+		std::vector<std::optional<YulName>> storageLoadFunctions;
+		std::vector<YulName> hashFunctions;
+	};
+	PredefinedBuiltinFunctions m_predefinedBuiltinFunctions;
+
+	size_t m_index {0};
+	size_t m_nGhosts {0};
+	std::vector<std::string> m_definedLabels {};
+	std::vector<std::tuple<YulName, bool>> m_names {};
```

Comment:
Shouldn’t we document the meaning of the false value in this context? Its purpose isn’t clear to me.

Reply:
It seems to be to know if a name is derived? But what is the whole idea on knowing that, if this is the case.

Reply:
The point is that you have a couple defined names which are known at a certain point in time (like variable names and such) and then the optimizer is going to add new names (eg when disambiguating). These new names do not need to have a string label yet but only when the AST (or parts of it) is printed into a human-readable format. For example:

```cpp
YulName name = repo.defineName("bla");  // this will yield some YulName / numerical id for "bla"
YulName name2 = repo.deriveName(name);  // this yields another id, derived from name but w/o label
YulName name3 = repo.deriveName(name3);  // this yields another id, derived from name2 but w/o label
repo.generateLabels();  // now the chains are cut and all derived names get their own label based on what they were derived from
cout << name2 << '\n';  // this could print, eg, "bla_1"
```